File: cpp/4.cpp
#include <iostream>
#include <cmath>
#include <string>
using namespace std;

bool isPrime (int a){ // проверка на простоту 
    if (a == 2) return true;
    if (a % 2 == 0) return false;
    for (int i = 3; i <= sqrt(a) + 1 ; i += 2){
        if (a % i == 0)  return false;
    }
    return true;
}

int modPow (int a, int b, int m){
    int res = 1;
    while (b > 0){
        if (b % 2 == 1) res = (res * a) % m;
        a = (a * a) % m;
        b /= 2;
    }
    return res;
}

string encrypt_message(const string& message, int key) {
    string encrypted_message = "";
    for (char c : message) {
        encrypted_message += static_cast<char>(c + key);
    }
    return encrypted_message;
}

string decrypt_message(const string& encrypted_message, int key) {
    string decrypted_message = "";
    for (char c : encrypted_message) {
        decrypted_message += static_cast<char>(c - key);
    }
    return decrypted_message;
}

int main () {
    system("chcp 65001");
    int secretAlice, secretBob, osnov, module;
    cout << "Введите секретный ключ Алисы: "; cin >> secretAlice;
    cout << "Введите секретный ключ Боба: "; cin >> secretBob;
    cout << "Введите основание: "; cin >> osnov;
    cout << "Введите модуль для формулы (простое число): "; cin >> module;
    if (!isPrime(module)){
        cout << "Модуль должен быть простым" << endl;
        return -1;
    }
    int publicAlice = modPow( osnov, secretAlice, module); // генерация публичного ключа (общее основание) ^ (секрет каждого) mod (общее простое число)
    int publicBob = modPow( osnov, secretBob, module);
    int keyAlice = modPow( publicBob, secretAlice, module); // итоговый общий ключ
    int keyBob = modPow( publicAlice, secretBob, module);// (полученный публичный ключ)^(секрет каждого ) mod (общее простое число)
    int key = 0;
    if (keyAlice == keyBob){
        key = keyAlice;
    }
    string message;
    cin.ignore(); // Очистка буфера ввода
    cout << "Введите сообщение для шифрования: ";
    getline(cin, message);


    string encrypted = encrypt_message(message, key);
    cout << "Зашифрованное сообщение: " << encrypted << endl;

    string decrypted = decrypt_message(encrypted, key);
    cout << "Расшифрованное сообщение: " << decrypted << endl;

    return 0;

}

File: go/4.go
package main

import (
	"fmt"
	"math"
)

func isPrime(a int) bool {
	if a == 2 {
		return true
	}
	if a%2 == 0 {
		return false
	}
	for i := 3; i <= int(math.Sqrt(float64(a)))+1; i += 2 {
		if a%i == 0 {
			return false
		}
	}
	return true
}

func modPow(a, b, m int) int {
	res := 1
	for b > 0 {
		if b%2 == 1 {
			res = (res * a) % m
		}
		a = (a * a) % m
		b /= 2
	}
	return res
}

func encryptMessage(message string, key int) string {
	encrypted := ""
	for _, c := range message {
		encrypted += string(c + rune(key))
	}
	return encrypted
}

func decryptMessage(encryptedMessage string, key int) string {
	decrypted := ""
	for _, c := range encryptedMessage {
		decrypted += string(c - rune(key))
	}
	return decrypted
}

func main() {
	var secretAlice, secretBob, osnov, module int
	fmt.Print("Введите секретный ключ Алисы: ")
	fmt.Scan(&secretAlice)
	fmt.Print("Введите секретный ключ Боба: ")
	fmt.Scan(&secretBob)
	fmt.Print("Введите основание: ")
	fmt.Scan(&osnov)
	fmt.Print("Введите модуль для формулы (простое число): ")
	fmt.Scan(&module)

	if !isPrime(module) {
		fmt.Println("Модуль должен быть простым")
		return
	}

	publicAlice := modPow(osnov, secretAlice, module)
	publicBob := modPow(osnov, secretBob, module)
	keyAlice := modPow(publicBob, secretAlice, module)
	keyBob := modPow(publicAlice, secretBob, module)

	if keyAlice != keyBob {
		fmt.Println("Ошибка: ключи не совпадают")
		return
	}

	key := keyAlice

	var message string
	fmt.Print("Введите сообщение для шифрования: ")
	fmt.Scanln(&message)

	encrypted := encryptMessage(message, key)
	fmt.Println("Зашифрованное сообщение:", encrypted)

	decrypted := decryptMessage(encrypted, key)
	fmt.Println("Расшифрованное сообщение:", decrypted)
}
